File: java/src/main/java/com/bookstore/AbstractItem.java
package com.bookstore;

/**
 * Base class of all our physical items. Most items are books but we also offer other things like
 * gift cards.
 */
public abstract class AbstractItem implements CartAble {

    private static final int NOT_READY_TO_DELIVER = -1;

    private final String name;
    private final DeliveryCostLines deliveryCostLines;
    private final DeliveryCostCalculator deliveryCostCalculator = new DeliveryCostCalculator();

    private boolean readyToDeliver;

    private final CartActions postAddCart;
    private final CartActions postPrepareDelivery;

    public AbstractItem(String name, DeliveryCostLines deliveryCostLines, CartActions postAddCart,
            CartActions postPrepareDelivery) {
        this.name = name;
        this.deliveryCostLines = deliveryCostLines;
        this.postAddCart = postAddCart;
        this.postPrepareDelivery = postPrepareDelivery;
    }

    public String getName() {
        return name;
    }

    @Override
    public void putIntoMyCart(Cart cart) {
        cart.add(getName());
        postAddCart.run(cart);
    }

    public void prepareDeliveryIn(Cart cart) {
        if (!cart.contains(this.name)) {
            throw new IllegalStateException("Can only calculate delivery costs for items in cart");
        }

        deliveryCostLines.forEach(costLine -> deliveryCostCalculator.apply(costLine, cart));
        markReadyToDeliver();
        postPrepareDelivery.run(cart);
    }

    protected final void markReadyToDeliver() {
        readyToDeliver = true;
    }

    public final int getDeliveryCost() {
        if (readyToDeliver) {
            return deliveryCostCalculator.calculate();
        }
        return NOT_READY_TO_DELIVER;
    }

    @Override
    public final String toString() {
        return name + " (" + getDeliveryCost() + "EUR)";
    }
}

File: python/Book-Discounts-Refactoring-Kata/bookstore/abstract_item.py
import math
from abc import abstractmethod

from bookstore.cart import Cart
from bookstore.cartable import CartAble


class AbstractItem(CartAble):
    """Base class of all our physical items. Most items are books, but we also offer other things like gift cards."""
    NOT_READY_TO_DELIVER: int = -1

    def __init__(self, name: str) -> None:
        self.name = name
        self.delivery_cost = 0
        self.ready_to_deliver = False

    @abstractmethod
    def item_weight(self) -> int:
        pass

    def prepare_delivery_in(self, cart: Cart):
        if not cart.contains(self.name):
            raise RuntimeError()

        grams: int = self.item_weight()
        self.calculate_delivery_cost(cart, grams)
        self.mark_ready_to_deliver()

    def calculate_delivery_cost(self, cart: Cart, grams: int):
        if grams <= 500:
            self.delivery_cost = 10
        else:
            self.delivery_cost = int(math.sqrt(grams))

    def mark_ready_to_deliver(self):
        self.ready_to_deliver = True

    def get_delivery_cost(self) -> int:
        return self.delivery_cost if self.ready_to_deliver else AbstractItem.NOT_READY_TO_DELIVER

    def __str__(self) -> str:
        return f"{self.name} ({self.get_delivery_cost()}EUR)"
